File: python/atelier_bench/rules/rush_surcharge.py
"""rush-surcharge rule (nominal: subclasses RuleContract)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from atelier_core import Priority

from ..contracts.rule_contract import RuleContract

if TYPE_CHECKING:
    from ..dataset import Dataset


class RushSurchargeRule(RuleContract):
    KEY = "rush-surcharge"

    def __init__(self) -> None:
        self.key = self.KEY

    def evaluate(self, data: "Dataset") -> bool:
        return any(order.priority is Priority.RUSH for order in data.orders)

File: go/atelier/rules/rush_surcharge.go
// Package rules carries the 48 rules: 24 NOMINAL rule structs in this package
// (each with a compile-time assertion against RuleContract and a value
// receiver) plus the 24 STRUCTURAL ad-hoc rule values from rules/adhoc, which
// assert nothing and satisfy the same method set by shape alone.
package rules

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	coresupport "atelier.example/lane/core/support"
)

// Compile-time proof against the NOMINAL contract, asserted on the VALUE
// type: RushSurchargeRule's methods take value receivers, so both RushSurchargeRule and
// *RushSurchargeRule satisfy RuleContract.
var _ contracts.RuleContract = RushSurchargeRule{}

// RushSurchargeRule is the rush-surcharge rule.
type RushSurchargeRule struct{}

// Key is the rule identifier.
func (RushSurchargeRule) Key() string {
	return "rush-surcharge"
}

// Evaluate reports whether the rule holds on this dataset.
func (r RushSurchargeRule) Evaluate(data *dataset.Dataset) bool {
	for _, order := range data.Orders {
		if order.Priority == coresupport.PriorityRush {
			return true
		}
	}
	return false
}

// init registers this rule.
func init() {
	Register(RushSurchargeRule{})
}
